Automattic/woocommerce-payments
# Fix infinite recursion when having more than 10 saved cards

#### Changes proposed in this Pull Request

* The `woocommerce_get_customer_payment_tokens` function bails if there are already 10 or more payment tokens in the array passed to the filter. That could indicate that there are more "pages" of tokens stored in the database, and in that case it's useless to try adding more.
* Remove the `woocommerce_get_customer_payment_tokens` filter before adding a token (and add it back again afterwards) so there's no useless recursion happening, because adding a token to the database will end up calling the `woocommerce_get_customer_payment_tokens` filter.

This is not the ideal solution, users with 11 saved cards won't be able to use the 11th one. However, I consider that to be an extremely uncommon edge case, and "failing" to store the 11th credit card seems like a much better experience than the checkout crashing. It doesn't seem worth the effort to try to properly fix this problem (I'm open to disagreements). Also, the real fix should come from WC Core, it doesn't make a lot of sense that the number of saved payment methods is tied to the `posts_per_page` option.

#### Testing instructions

- Double-check that `SELECT option_value FROM wp_options WHERE option_name="post_per_page"` is 10 (it's the default).
- To remove some noise, run `TRUNCATE wp_woocommerce_payment_tokens;` and `TRUNCATE wp_woocommerce_payment_tokenmeta;`
- Add saved cards to a Customer until it has more than 10 (or whatever `posts_per_page` is in your test site).
- Go to the Stripe Dashboard (the Platform dashboard, not the merchant-accessible one).
- Logged-in as that user, go to the Checkout.
- See that the page doesn't crash. Take a moment to celebrate.
- Confirm that you see 10 saved cards in the checkout.
- Go to the database, confirm that, no matter how many times you refresh the checkout, the amount of rows in `wp_woocommerce_payment_tokens` doesn't grow.

This fix (or a different version of it) should also be applied to the Stripe extension (props @luizreis for pointing that out).
